### PR TITLE
fix: handle undefined destination in transformation response

### PR DESCRIPTION
**Pull request by MoumitaM**
## PR Description

Handled undefined destination in transformation response

## Notion ticket

Ticket link

## Screenshots

Please add screenshots for any new features or UI bug fixes for the following browsers -

- Chrome
  >
- Firefox
  >
- Safari
  >

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: src/core/analytics.js [hunk 1]
```diff
@@ -392,16 +392,19 @@ class Analytics {
                 let transformedEvents;
                 if (transformationServerAccess) {
                   // filter the transformed event for that destination
-                  transformedEvents = transformedPayload.find(
+                  const filteredDestination = transformedPayload.find(
                     (e) => e.id === intg.destinationId,
-                  ).payload;
+                  );
+                  transformedEvents = filteredDestination ? filteredDestination.payload : undefined;
```
Comment (saikumarrs) [inline src/core/analytics.js:398]:
```suggestion
                  transformedEvents = transformedPayload.find(
                    (e) => e.id === intg.destinationId,
                  )?.payload;
```

File: src/core/analytics.js [hunk 2]
```diff
@@ -392,16 +392,19 @@ class Analytics {
...
-                // send transformed event to destination
-                transformedEvents.forEach((tEvent) => {
-                  this.sendDataToDestination(intg, { message: tEvent.event }, methodName);
-                });
+                if (transformedEvents) {
+                  // send transformed event to destination
+                  transformedEvents.forEach((tEvent) => {
+                    this.sendDataToDestination(intg, { message: tEvent.event }, methodName);
+                  });
+                }
```
Comment (saikumarrs) [inline src/core/analytics.js:407]:
```suggestion
                // send transformed event to destination
                transformedEvents?.forEach((tEvent) => {
                  this.sendDataToDestination(intg, { message: tEvent.event }, methodName);
                });
```

**Review by saikumarrs — APPROVED** [review]

File: src/core/analytics.js (hunk 1 above)
Comment (MoumitaM) [reply src/core/analytics.js:398]: This is `analytics.js` file, AFAIR this doesn't work.

File: src/core/analytics.js (hunk 2 above)
Comment (MoumitaM) [reply src/core/analytics.js:407]: This is `analytics.js` file, AFAIR this doesn't work.

File: src/core/analytics.js (hunk 1 above)
Comment (MoumitaM) [reply src/core/analytics.js]: Done

File: src/core/analytics.js (hunk 2 above)
Comment (MoumitaM) [reply src/core/analytics.js]: Updated logic

Comment (saikumarrs) [reply src/core/analytics.js]: Is optional chaining not working?

Comment (MoumitaM) [reply src/core/analytics.js]: Done

File: src/features/core/deviceModeTransformation/transformationHandler.js [hunk 3]
```diff
@@ -83,19 +83,8 @@ class TransformationsHandler {
...
                     } 
                 */
-                /**
-                 * Filter the successful transformed event
-                 */
-                const transformationResponse = [];
-                response.transformedBatch.forEach((dest) => {
-                  transformationResponse.push({
-                    id: dest.id,
-                    payload: dest.payload.filter((tEvent) => tEvent.status === '200'),
```
Comment (bardisg) [inline src/features/core/deviceModeTransformation/transformationHandler.js:93]: why we took this one off? according to DMT developers we should be getting errored response objects too in the array. should we not sent only 200 status with not null | undefined value for event property?

File: src/core/analytics.js [hunk 4]
```diff
@@ -392,15 +392,26 @@ class Analytics {
...
+                  transformedEvents = transformedEvents?.payload.filter(
+                    (tEvent) => tEvent.status === '200',
+                  );
                 } else {
                   transformedEvents = transformedPayload;
                 }
                 // send transformed event to destination
-                transformedEvents.forEach((tEvent) => {
-                  this.sendDataToDestination(intg, { message: tEvent.event }, methodName);
+                transformedEvents?.forEach((tEvent) => {
```
Comment (bardisg) [inline src/core/analytics.js:411]: should we not sent only 200 status with not null | undefined value for event property?

**Review by bardisg — CHANGES_REQUESTED** [review]

File: src/features/core/deviceModeTransformation/transformationHandler.js (hunk 3 above)
Comment (MoumitaM) [reply src/features/core/deviceModeTransformation/transformationHandler.js:93]: I have moved the code from here to analytics.js so the we can better handle the error message.

File: src/core/analytics.js (hunk 4 above)
Comment (MoumitaM) [reply src/core/analytics.js:411]: If the rudder event is null , there is no point sending it to device mode destination

**Review by bardisg — DISMISSED** [review]

File: src/core/analytics.js [hunk 5]
```diff
@@ -392,15 +392,27 @@ class Analytics {
                 let transformedEvents;
```
Comment (saikumarrs) [inline src/core/analytics.js:392]: Refactored to handle some failure scenarios.
```suggestion
              try {
                let transformedEvents = [];
                if (transformationServerAccess) {
                  // filter the transformed event for that destination
                  const destTransformedResult = transformedPayload.find((e) => e.id === intg.destinationId);
                  if (!destTransformedResult) {
                    logger.error(
                      `[DMT]::Transformed data for destination "${intg.name}" was not sent from the server`,
                    );
                    return;
                  }

                  destTransformedResult?.payload.forEach((tEvent) => {
                    if (tEvent.status === '200') {
                      transformedEvents.push(tEvent);
                    } else {
                      logger.error(
                        `[DMT]::Event transformation unsuccessful for destination "${intg.name}". Dropping the event. Status: "${tEvent.status}". Error Message: "${tEvent.error}"`,
                      );
                    }
                  });
                } else {
                  transformedEvents = transformedPayload;
                }
                // send transformed event to destination
                transformedEvents?.forEach((tEvent) => {
                  if (tEvent.event) {
                    this.sendDataToDestination(intg, { message: tEvent.event }, methodName);
                  }
                });
              } catch (e) {
                if (e instanceof Error) {
                  e.message = `[DMT]::[Destination:${intg.name}]:: ${e.message}`;
                }
                handleError(e);
              }
```

**Review by saikumarrs — COMMENTED** [review]
Refactor.

**Review by saikumarrs — APPROVED** [review]

**Review by bardisg — APPROVED** [review]

**Pull request merged**